Sources/Apollo/ApolloStore.swift

```diff
@@ -99,14 +102,32 @@ public final class ApolloStore {
       self.cacheLock.unlock()
     }
   }
-
-  public func withinReadTransaction<T>(_ body: @escaping (ReadTransaction) throws -> T) -> Promise<T> {
-    return withinReadTransaction {
-      Promise(fulfilled: try body($0))
+  
+  /// Performs an operation within a read transaction
+  ///
+  /// - Parameters:
+  ///   - body: The body of the operation to perform.
+  ///   - callbackQueue: [optional] The callback queue to use to perform the completion block on. Will perform on the current queue if not provided. Defaults to nil.
+  ///   - completion: [optional] The completion block to perform when the read transaction completes. Defaults to nil.
+  public func withinReadTransaction<T>(_ body: @escaping (ReadTransaction) throws -> T,
+                                       callbackQueue: DispatchQueue? = nil,
+                                       completion: ((Result<T, Error>) -> Void)? = nil) {
+    _ = self.withinReadTransactionPromise {
```

Comment:
This may not be worth it, but maybe a generic wrapper method like [Node's `callbackify`](https://nodejs.org/api/util.html#util_util_callbackify_original) could be used to abstract this conversion.